home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/CooperationServiceImpl.java

```diff
@@ -0,0 +1,103 @@
+        cooperation.getHouses().forEach(element -> {
+            element.setCooperation(cooperation);
+            element.setCreateDate(LocalDateTime.now());
+            element.setEnabled(true);
+        });
+        cooperationRepository.save(cooperation);
+        return mapper.convert(cooperation, CooperationDto.class);
+    }
+
+    @Transactional
+    @Override
+    public CooperationDto updateCooperation(Long id, CooperationDto updateCooperationDto) {
+        Optional<Cooperation> optional = cooperationRepository.findById(id).filter(Cooperation::getEnabled);
+        if (optional.isPresent()) {
+            Cooperation fromDb = optional.get();
+
+            if (updateCooperationDto.getName() != null) {
+                fromDb.setName(updateCooperationDto.getName());
+            }
+            if (updateCooperationDto.getUsreo() != null) {
+                fromDb.setUsreo(updateCooperationDto.getUsreo());
+            }
+            if (updateCooperationDto.getIban() != null) {
+                fromDb.setIban(updateCooperationDto.getIban());
+            }
+            if (updateCooperationDto.getAddress() != null) {
+                fromDb.setAddress(updateCooperationDto.getAddress());
+            }
+            fromDb.setUpdateDate(LocalDateTime.now());
+            cooperationRepository.save(fromDb);
+            return mapper.convert(fromDb, CooperationDto.class);
+        } else {
+            throw new NotFoundHomeException(String.format(NOT_FOUND_COOPERATION_FORMAT, id));
+        }
+    }
+
+    @Transactional
+    @Override
+    public Page<CooperationDto> getAllCooperation(Integer pageNumber, Integer pageSize,
+                                                  Specification<Cooperation> specification) {
+        Specification<Cooperation> cooperationSpecification = specification
+            .and((root, criteriaQuery, criteriaBuilder) -> criteriaBuilder.equal(root.get("enabled"), true));
+        return cooperationRepository.findAll(cooperationSpecification, PageRequest.of(pageNumber - 1, pageSize))
+            .map(cooperation -> mapper.convert(cooperation, CooperationDto.class));
+    }
+
+    @Override
+    public CooperationDto getCooperationById(Long id) {
+        Cooperation toGet = cooperationRepository.findById(id).filter(Cooperation::getEnabled)
+            .orElseThrow(() -> new NotFoundHomeException(String.format(NOT_FOUND_COOPERATION_FORMAT, id)));
+        List<House> houseList = houseRepository.findHousesByCooperationId(toGet.getId());
+        toGet.setHouses(houseList);
+        return mapper.convert(toGet, CooperationDto.class);
+    }
+
+    @Override
+    public void deactivateCooperation(Long id) {
+        Cooperation toDelete = cooperationRepository.findById(id).filter(Cooperation::getEnabled)
+            .orElseThrow(() -> new NotFoundHomeException(String.format(NOT_FOUND_COOPERATION_FORMAT, id)));
+        toDelete.setEnabled(false);
+        cooperationRepository.save(toDelete);
```

Comment:
Good question. It depends on how we are going to use children after disabling parents. To avoid huge cascade updates let's disable the only parents and in the future keep the children invisible in the system if a parent is disabled.